x-pack/plugins/reporting/server/lib/content_stream.ts

```diff
@@ -249,8 +251,37 @@ export class ContentStream extends Duplex {
     });
   }
 
-  private async flush(size = this.buffer.byteLength) {
-    const chunk = this.buffer.slice(0, size);
+  private async flush(size = this.bytesBuffered) {
+    const buffersToFlush: Buffer[] = [];
+    let totalBytesConsumed = 0;
+
+    /*
+     Loop over each buffer, keeping track of how many bytes we have added
+     to the array of buffers to be flushed. The array of buffers to be flushed
+     contains buffers by reference, not copies. This avoids putting pressure on
+     the CPU for copying buffers or for gc activity. Please profile performance
+     with a large byte configuration and a large number of records (900k+)
+     before changing this code. Config used at time of writing:
+
+      xpack.reporting:
+        csv.maxSizeBytes: 500000000
+        csv.scroll.size: 1000
+
+     At the moment this can put memory pressure on Kibana. Up to 1,1 GB in a dev
+     build. It is not recommended to have overly large max size bytes but we
+     need this code to be as performant as possible.
+    */
+    for (const buffer of this.buffers) {
+      if (totalBytesConsumed + buffer.byteLength <= size) {
+        buffersToFlush.push(buffer);
+        totalBytesConsumed += buffer.byteLength;
+      } else {
+        break;
+      }
```

Comment:
Could you please invert the condition to improve the readability of this block?

```suggestion
      if (totalBytesConsumed + buffer.byteLength > size) {
        break;
      }

      buffersToFlush.push(buffer);
      totalBytesConsumed += buffer.byteLength;
```